velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -731,6 +968,108 @@ static bool compareMaterializedRows(
   return true;
 }
 
+bool assertEqualResults(
+    const std::vector<RowVectorPtr>& expected,
+    const std::vector<RowVectorPtr>& actual) {
+  MaterializedRowMultiset expectedRows;
+  for (auto vector : expected) {
+    auto rows = materialize(vector);
+    std::copy(
+        rows.begin(),
+        rows.end(),
+        std::inserter(expectedRows, expectedRows.end()));
+  }
+
+  return assertEqualResults(expectedRows, actual);
+}
+
+/// Returns the number of floating-point columns and a list of columns indices
+/// with floating-point columns placed at the end.
+std::tuple<uint32_t, std::vector<velox::column_index_t>>
+findFloatingPointColumns(const MaterializedRow& row) {
+  auto isFloatingPointColumn = [&](size_t i) {
+    return row[i].kind() == TypeKind::REAL || row[i].kind() == TypeKind::DOUBLE;
+  };
+
+  uint32_t numFloatingPointColumns = 0;
+  std::vector<velox::column_index_t> indices;
+  for (auto i = 0; i < row.size(); ++i) {
+    if (isFloatingPointColumn(i)) {
+      ++numFloatingPointColumns;
+    } else {
+      indices.push_back(i);
+    }
+  }
+
+  for (auto i = 0; i < row.size(); ++i) {
+    if (isFloatingPointColumn(i)) {
+      indices.push_back(i);
+    }
+  }
+  return std::make_tuple(numFloatingPointColumns, indices);
+}
+
+// Compare actualRows with expectedRows and return whether they match. Compare
+// actualRows and expectedRows with epsilon if needed and allowed. Otherwise,
+// compare their values directly. The underlying assumption is that aggregation
+// results can be sorted by unique keys and floating-point values in them are
+// computed in different ways and hence require epsilon comparison. For results
+// of other operations, floating-point values are likely copied from inputs and
+// hence can be compared directly.
+bool assertEqualResults(
+    const MaterializedRowMultiset& expectedRows,
+    const MaterializedRowMultiset& actualRows,
+    const std::string& message) {
+  if (expectedRows.empty() != actualRows.empty()) {
+    LOG(ERROR) << generateUserFriendlyDiff(expectedRows, actualRows) << message;
```

Comment:
Why do we log here and not use EXPECT_TRUE(false) << ...error message...? Same question for other logging statements.

Reply:
It's because I'd like to test unmatched comparisons in the unit test. If we use EXPECT_TRUE(false) here, the unit test always fail.

Reply:
But doesn't that mean that the actual test won't fail. Do we need to ensure that all callers assert on the result. For example, the following code will not fail the test on failure, only log an error. Is this intentional?

```
  assertEqualResults(
      expectedRows,
      actualRows,
      kDuckDbTimestampWarning + "\nDuckDB query: " + duckDbSql);
```

Reply:
Maybe there is a way to use a test listener to "catch" test failures in a test.

https://google.github.io/googletest/advanced.html#propagating-fatal-failures

Reply:
> But doesn't that mean that the actual test won't fail. Do we need to ensure that all callers assert on the result. For example, the following code will not fail the test on failure, only log an error. Is this intentional?
> 
> ```
>   assertEqualResults(
>       expectedRows,
>       actualRows,
>       kDuckDbTimestampWarning + "\nDuckDB query: " + duckDbSql);
> ```

Good catch! There are two callsites of assertEqualResults() that should assert the returned value is true. I'll fix them.